src/main/webapp/resources/js/pages/projects/samples-metadata-import/redux/importReducer.ts

```diff
@@ -0,0 +1,353 @@
+}
+
+export interface MetadataSaveDetailsItem {
+  saved: boolean;
+  error?: string;
+}
+
+export interface SaveMetadataResponse {
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+export interface SetSampleNameColumnResponse {
+  sampleNameColumn: string;
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+}
+
+export interface InitialState {
+  projectId: string;
+  sampleNameColumn: string;
+  headers: MetadataHeaderItem[];
+  metadata: MetadataItem[];
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+  percentComplete: number;
+}
+
+const initialState: InitialState = {
+  projectId: "",
+  sampleNameColumn: "",
+  headers: [],
+  metadata: [],
+  metadataValidateDetails: {},
+  metadataSaveDetails: {},
+  percentComplete: 0,
+};
+
+/*
+Redux async thunk for saving the metadata to samples.
+For more information on redux async thunks see: https://redux-toolkit.js.org/api/createAsyncThunk
+ */
+export const saveMetadata = createAsyncThunk<
+  SaveMetadataResponse,
+  { projectId: string; selectedMetadataKeys: string[] },
+  { dispatch: ImportDispatch; state: ImportState; rejectValue: string }
+>(
+  `importReducer/saveMetadata`,
+  async (
+    { projectId, selectedMetadataKeys },
+    { dispatch, getState, rejectWithValue }
+  ) => {
+    const state: ImportState = getState();
+    const {
+      sampleNameColumn,
+      headers,
+      metadata,
+      metadataValidateDetails,
+      metadataSaveDetails,
+    } = state.importReducer;
+
+    try {
+      //save header details (metadata field & restriction)
+      //if failure display error notification on page
+      await createMetadataFieldsForProject({
+        projectId,
+        body: headers
+          .filter((header) => header.name !== sampleNameColumn)
+          .map((header) => ({
+            label: header.name,
+            restriction: header.targetRestriction,
+          })),
+      }).catch((error) => {
+        throw new Error(error.response.data.error);
+      });
+
+      //save selected metadata entry rows
+      //during a partial failure only save data that has not already been saved
+      const selectedSampleList = metadata.filter((metadataItem) => {
+        const name: string = metadataItem[sampleNameColumn];
+        return (
+          selectedMetadataKeys.includes(metadataItem.rowKey) &&
+          metadataSaveDetails[name]?.saved !== true
+        );
+      });
+
+      const createSampleList: CreateSampleItem[] = [];
+      const updateSampleList: UpdateSampleItem[] = [];
+
+      selectedSampleList.forEach((metadataItem) => {
+        const name = metadataItem[sampleNameColumn];
+        const sampleId = metadataValidateDetails[name].foundSampleId;
+        const metadata = createMetadataFields(
+          sampleNameColumn,
+          headers,
+          metadataItem
+        );
+
+        if (typeof sampleId === "number") {
```

Comment:
Can you check your logic here, I am getting the following editor warning: "Redundant 'typeof' check: 'sampleId' always has type 'number' "

Reply:
Please see 40ca2224aaaa010ae59fc99e824f9be8fee32cc7.

Reply:
That fixed the warning I was getting, thank you.